lib/execution_engine2/utils/clients.py

```diff
@@ -0,0 +1,112 @@
+            raise ValueError("token is required")
+        if not cfg:
+            raise ValueError("cfg is required")
+        self._user_id = user_id
+        self._token = token
+
+        # Do a check that the url actually points to the workspace?
+        # Also maybe consider passing in the workspace url rather than the dict, but the ClientSet
+        # below will need lots of params so a dict makes sense there, maybe keep the apis similar?
+        # TODO the client throws a 'X is not a valid url' error if the url isn't valid, improve
+        #      by catching & rethrowing with a more clear message that the config is wrong
+        ws_url = cfg.get("workspace-url")  # may want to make the keys constants?
+        if not ws_url or not ws_url.strip():
+            raise ValueError("missing workspace-url in configuration")
+        self._workspace = Workspace(ws_url, token=token)
+        self._workspace_auth = WorkspaceAuth(user_id, self._workspace)
+
+    # create_autospec can't mock instance variables, so we add some boilerplate
+    # Could make these properties but then they return MagicMocks which don't update with API
+    # changes
+
+    def user_id(self):
+        return self._user_id
+
+    def token(self):
+        return self._token
+
+    def workspace(self):
+        return self._workspace
+
+    def workspace_auth(self):
+        return self._workspace_auth
+
+
+class ClientSet:
+    """
+    Clients required by EE2 for communicating with other services.
+
+    These are not user-specific and can be reused throughout the application.
+    """
+
+    def __init__(self, cfg: Dict[str, str], cfg_path: str, debug: bool = False):
+        """
+        Initialize the client set from a configuration dictionary.
+
+        cfg - the configuration dictionary
+        cfg_path - the path to the configuration file
+        debug - set clients that support it to debug mode
+
+        Expected keys in config:
+        auth-url - the root URL of the kbase auth service
+        catalog-url - the URL of the catalog service
+        catalog-token - a token to use with the catalog service. Ideally a service token
+        kafka-host - the host string for a Kafka service
+        slack-token - a token for contacting Slack
+        """
+        # TODO seems like it'd make sense to init Condor from a config dict like everything else
+        self.condor = Condor(cfg_path)
+        self.catalog_utils = CatalogUtils(cfg["catalog-url"], cfg["catalog-token"])
+        auth_url = cfg["auth-url"]
+        self.auth = KBaseAuth(auth_url=auth_url + "/api/legacy/KBase/Sessions/Login")
+        # TODO using hardcoded roles for now to avoid possible bugs with mismatched cfg roles
+        #       these should probably be configurable
+        self.auth_admin = AdminAuthUtil(auth_url, [ADMIN_READ_ROLE, ADMIN_WRITE_ROLE])
+
+        # KafkaClient has a nice error message when the arg is None
+        self.kafka_client = KafkaClient(cfg.get("kafka-host"))
```

Comment:
Maybe we should follow the fail fast with Keyerror way to handling dictionaries by doing the following
```
cfg['key'] == required value
cfg.get('key') == optionally None
cfg.get('key','default') == explicitly set default
```

Reply:
If we wanted to do something like this the way it'd do it is something like:
```
if not cfg.get('key') or not cfg.get('key').strip():
    raise ValueError('configuration is missing key "key"')
```
Just getting a KeyError("key") is not very helpful to someone trying to figure out why the service won't start, especially if they're not familiar with it.

I actually did use the first approach in cases where the client I was instantiating ignored missing arguments, although there could still be issues if the argument was whitespace or a bad url, etc. Where the clients had readable error messages (like Kafka and Slack) I went with those.

In most of my repos all the config args are checked for correctness, and urls used to make requests to the service and check the results, before the service starts. The checking is mostly delegated to the clients / objects that use them since they understand the semantics.

Reply:
Add dict['key'] vs dict.get('') to the style guide